app/src/main/java/com/nutomic/syncthingandroid/model/Completion.java

```diff
@@ -0,0 +1,175 @@
+package com.nutomic.syncthingandroid.model;
+
+import android.util.Log;
+
+import com.nutomic.syncthingandroid.util.DefaultHashMap;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+
+/**
+ * This class caches remote folder and device synchronization
+ * completion indicators defined in {@link CompletionInfo#CompletionInfo}
+ * according to syncthing's REST "/completion" JSON result schema.
+ * Completion model of syncthing's web UI is completion[deviceId][folderId]
+ */
+public class Completion {
+
+    private static final String TAG = "Completion";
+
+    DefaultHashMap<String, DefaultHashMap<String, CompletionInfo>> deviceFolderMap =
+        new DefaultHashMap<String, DefaultHashMap<String, CompletionInfo>>();
+
+    DefaultHashMap<String, CompletionInfo> defaultDevice = new DefaultHashMap<String, CompletionInfo>();
+    CompletionInfo defaultFolder = new CompletionInfo();
+
+    /**
+     * Adds a device to the cache model if it does not exist.
+     */
+    /*
+    private void addDevice(String deviceId) {
+        if (!deviceFolderMap.containsKey(deviceId)) {
+            deviceFolderMap.put(deviceId, defaultDevice);
+        }
+    }
+    */
+
+    /**
+     * Removes a device from the cache model.
+     */
+    /*
+    private void removeDevice(String deviceId) {
+        if (deviceFolderMap.containsKey(deviceId)) {
+            deviceFolderMap.remove(deviceId);
+        }
+    }
+    */
+
+    /**
+     * Adds a folder to the cache model if it does not exist.
+     */
+    private void addFolder(String deviceId, String folderId) {
+        // Add device parent node if it does not exist.
+        if (!deviceFolderMap.containsKey(deviceId)) {
+            deviceFolderMap.put(deviceId, defaultDevice);
+        }
+        // Add folder.
+        if (!deviceFolderMap.get(deviceId).containsKey(folderId)) {
+            deviceFolderMap.get(deviceId).put(folderId, defaultFolder);

```

Comment:
All devices all folders will refer to the same object instance which I think is not what you want?

Reply:
Yes, I reverted it to the new <object> code I had before.